factor out conntrack GC into separate package

ianvernon opened the pull request:
CT garbage collection uses the EndpointManager API - it does not need to be
tied into `pkg/endpointmanager`. Remove it accordingly.

Signed-off by: Ian Vernon <ian@cilium.io>

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/9160)
<!-- Reviewable:end -->

Commit:
factor out conntrack GC into separate package

CT garbage collection uses the EndpointManager API - it does not need to be
tied into `pkg/endpointmanager`. Remove it accordingly.

Signed-off by: Ian Vernon <ian@cilium.io>

aanm commented on pkg/maps/ctmap/gc/gc.go:
```diff
@@ -21,12 +21,55 @@ import (
+var log = logging.DefaultLogger.WithField(logfields.LogSubsys, "ct-gc")
+
+// EnableConntrackGC enables the connection tracking garbage collection.
+func EnableConntrackGC(ipv4, ipv6 bool, restoredEndpoints []*endpoint.Endpoint, mgr *endpointmanager.EndpointManager) {
```
Don't import the endpointmanager directly, use an interface instead to avoid cycles in the future. (By extension we could almost do the same for the `endpoint.Endpoint`. Feel free to open a GH issue good-first-issue for this last one)

ianvernon replied:
Will add an interface for `EndpointManager`. The last one is a bit more difficult because `GetEndpoints` returns a slice of Endpoint pointers; we can't have the return type for `GetEndpoints` not be a slice of `*endpoint.Endpoint`, because that means that `GetEndpoints` won't actually implement that interface :/ .

ianvernon replied:
Added the interface per your request, PTAL

joestringer reviewed (changes requested):
From a BPF interaction perspective this is completely fine; it's just moving code from one package to another. So I'm fine with that.

Looking at the new package though I'm not sure why a `pkg/maps/ctmap/gc` package inherently needs to interact with the endpointmanager interface. Ideally what I'd look for is this package (or maybe even just the `pkg/maps/ctmap`) has a `GC()` function, then the iteration over endpoints in the endpointmanager is handled from the endpointmanager. To achieve the same behaviour as today, we would need to make sure that there's a way to ensure that the global map is only cleaned once, while the endpoint-specific maps are all cleaned if the endpoint is configured to have local conntrack.

This proposal would properly separate the BPF dependency of map cleanup from the endpointmanager, assuming you have some way to inject the conntrack GC implementation into the endpointmanager.

What do you think?

joestringer reviewed (approved):
I'm fine with this as-is for the goal of removing the `ctmap` dependency from `endpointmanager`, even though logically I think this should end up being reworked a bit eventually. That said, without some notion of an endpoint's datapath there isn't an obvious location to place this logic just yet. So LGTM.

aanm commented on daemon/daemon_main.go:
```diff
@@ -1194,8 +1195,8 @@ func runDaemon() {
 	log.Info("Starting connection tracking garbage collector")
-	d.endpointManager.EnableConntrackGC(option.Config.EnableIPv4, option.Config.EnableIPv6,
-		restoredEndpoints.restored)
+	gc.EnableConntrackGC(option.Config.EnableIPv4, option.Config.EnableIPv6,
```
Small nit, this can now be called `gc.Enable(...)`

Commit:
gc: rename `EnableConntrackGC` to `GC`

The old naming is now redundant, make it more pithy.

Signed-off by: Ian Vernon <ian@cilium.io>
